software/controller/lib/debug/debug.cpp

```diff
@@ -177,170 +200,162 @@ bool DebugSerial::ReadNextByte() {
+  // termination character and start waiting on the next
   // command.
-  char tmp = static_cast<char>(DbgSpecial::TERM);
-  (void)Hal.debugWrite(&tmp, 1);
+  char end_transfer = static_cast<char>(SpecialChar::kEndTransfer);
+  (void)Hal.debugWrite(&end_transfer, 1);
 
-  poll_state_ = DbgPollState::WAIT_CMD;
+  state_ = State::kWait;
   response_bytes_sent_ = 0;
   return false;
 }
 
 // Process the received command
-void DebugSerial::ProcessCmd() {
-  // The total number of bytes received (not including
-  // the termination byte) is the value of request_size_.
-  // This should be at least 3 (command & checksum).
-  // If its not, I just ignore the command and jump
-  // to waiting on the next.
-  // This means we can send TERM characters to synchronize
-  // communications if necessary
+void SerialHandler::ProcessCmd() {
+  // The total number of bytes received (not including the termination byte)
+  // is the value of request_size_, which should be at least 3 (8 bits command
+  // code + 16 bits checksum). If its not, I just ignore the command and jump to
+  // waiting for the next one.
+  // This means we can send kEndTransfer characters to synchronize
+  // communication if necessary
   if (request_size_ < 3) {
     request_size_ = 0;
-    poll_state_ = DbgPollState::WAIT_CMD;
+    state_ = State::kWait;
     return;
   }
 
-  uint16_t crc = CalcCRC(request_, request_size_ - 2);
+  uint16_t crc = ComputeCRC(request_, request_size_ - 2);
 
   if (crc != u8_to_u16(&request_[request_size_ - 2])) {
-    SendError(DbgErrCode::CRC_ERR);
+    SendError(ErrorCode::kCrcError);
     return;
   }
 
-  DebugCmd *cmd = DebugCmd::cmd_registry_[request_[0]];
-  if (!cmd) {
-    SendError(DbgErrCode::BAD_CMD);
+  Command::Handler *cmd_handler = registry_[request_[0]];
+  if (!cmd_handler) {
+    SendError(ErrorCode::kUnknownCommand);
     return;
   }
 
   // The length that we pass in to the command handler doesn't
   // include the command code or CRC.  The max size is also reduced
   // by 3 to make sure we can add the error code and CRC.
-  int len_in = request_size_ - 3;
-  int len_out = 0;
-  CmdContext context = {.req = &request_[1],
-                        .req_len = len_in,
-                        .resp = &response_[1],
-                        .max_resp_len = static_cast<int>(sizeof(response_) - 3),
-                        .resp_len = 0};
-  DbgErrCode err = cmd->HandleCmd(&context);
-  if (err != DbgErrCode::OK) {
-    SendError(err);
+  Command::Context context = {
+      .request = &request_[1],
+      .request_length = request_size_ - 3,
+      .response = &response_[1],
+      .max_response_length = sizeof(response_) - 3,
+      .response_length = 0,
+  };
+  ErrorCode error = cmd_handler->Process(&context);
+  if (error != ErrorCode::kNone) {
+    SendError(error);
     return;
   }
 
-  len_out = context.resp_len;
-
-  response_[0] = static_cast<uint8_t>(DbgErrCode::OK);
+  response_[0] = static_cast<uint8_t>(ErrorCode::kNone);
```

Comment:
The rest of the function from this point on resembles a lot the contents of the `SendError` function. Any potential for factoring this out?

Reply:
sure thing